presto-geospatial/src/main/java/com/facebook/presto/geospatial/BingTileFunctions.java

```diff
@@ -174,20 +177,129 @@ public static Block geometryToBingTiles(@SqlType(GEOMETRY_TYPE_NAME) Slice input
+                writeTilesToBlockBuilder(geometry, zoomLevel, tile, blockBuilder);
+            }
+        }
 
         return blockBuilder.build();
     }
 
+    private static BingTile[] getTilesInBetween(BingTile leftUpperTile, BingTile rightLowerTile, int zoomLevel)
+    {
+        checkArgument(leftUpperTile.getZoomLevel() == rightLowerTile.getZoomLevel());
+        checkArgument(leftUpperTile.getZoomLevel() > zoomLevel);
+
+        int divisor = 1 << (leftUpperTile.getZoomLevel() - zoomLevel);
+        int minX = (int) Math.floor(leftUpperTile.getX() / divisor);
+        int maxX = (int) Math.floor(rightLowerTile.getX() / divisor);
+        int minY = (int) Math.floor(leftUpperTile.getY() / divisor);
+        int maxY = (int) Math.floor(rightLowerTile.getY() / divisor);
+
+        BingTile[] tiles = new BingTile[(maxX - minX + 1) * (maxY - minY + 1)];
+        int index = 0;
+        for (int x = minX; x <= maxX; x++) {
+            for (int y = minY; y <= maxY; y++) {
+                tiles[index] = BingTile.fromCoordinates(x, y, OPTIMIZED_TILING_MIN_ZOOM_LEVEL);
+                index++;
+            }
+        }
+
+        return tiles;
+    }
+
+    /**
+     * Identifies a minimum set of tiles at specified zoom level that cover intersection of the
+     * specified geometry and a specified tile of the same or lower level. Adds tiles to provided
+     * BlockBuilder.
+     */
+    private static void writeTilesToBlockBuilder(OGCGeometry geometry,
+                                                 int zoomLevel,
+                                                 BingTile tile,
+                                                 BlockBuilder blockBuilder)
+    {
+        int tileZoomLevel = tile.getZoomLevel();
+        checkArgument(tile.getZoomLevel() <= zoomLevel);
+
+        OGCGeometry polygon = tileToPolygon(tile);
+        if (tileZoomLevel == zoomLevel) {
+            if (!geometry.disjoint(polygon)) {
+                BIGINT.writeLong(blockBuilder, tile.encode());
+            }
+            return;
+        }
+
+        if (geometry.contains(polygon)) {
+            int subTileCount = 1 << (zoomLevel - tileZoomLevel);
+            int minX = subTileCount * tile.getX();
+            int minY = subTileCount * tile.getY();
+            for (int x = minX; x < minX + subTileCount; x++) {
+                for (int y = minY; y < minY + subTileCount; y++) {
+                    BIGINT.writeLong(blockBuilder, BingTile.fromCoordinates(x, y, zoomLevel).encode());
+                }
+            }
+            return;
+        }
+
+        if (geometry.disjoint(polygon)) {
+            return;
+        }
+
+        int minX = 2 * tile.getX();
+        int minY = 2 * tile.getY();
+        int nextZoomLevel = tileZoomLevel + 1;
+        for (int x = minX; x < minX + 2; x++) {
+            for (int y = minY; y < minY + 2; y++) {
+                writeTilesToBlockBuilder(
+                            geometry,
```

Comment:
indentation seems off here for these arguments too, can you please double check?